eng/pipelines/templates/steps/build-artifacts.yml

```diff
@@ -83,13 +90,42 @@ steps:
 
   - ${{ parameters.BeforePublishSteps }}
 
-  - publish: $(Build.ArtifactStagingDirectory)
-    artifact: packages 
-    condition: succeededOrFailed()
+  - pwsh: |
+      # handle where the parameter is undefined
+      $shouldPublishDocs = $true
+      $docParam = "${{ parameters.BuildDocs }}"
+      if ($docParam) {
+        # if we set it to false handle it here
+        $shouldPublishDocs = 	[System.Convert]::ToBoolean($docParam)
+      }
+      
+      $packagePublishes = ((Get-ChildItem '$(Build.ArtifactStagingDirectory)') | Measure-Object).Count
+      
+      if ($packagePublishes -eq 0) {
+          Write-Host "We are expecting to publish package artifacts, but no artifacts were found."
+          Write-Host (Get-ChildItem '$(Build.ArtifactStagingDirectory)')
+          exit 1
+      }
+
+      if ($shouldPublishDocs) {
+        $docsPublishes = ((Get-ChildItem '$(Build.SourcesDirectory)/_docs') | Measure-Object).Count
+        if ($docsPublishes -eq 0) {
+          Write-Host "We are expecting to publish documentation artifacts, but no documentation artifacts were found."
+          Write-Host (Get-ChildItem '$(Build.SourcesDirectory)/_docs')
+          exit 1
+        }
+      }
+    displayName: 'Verify Presence of Artifacts'
 
   - task: PublishBuildArtifacts@1
-    condition: and(succeededOrFailed(), ${{parameters.BuildDocs}})
```

Comment:
Wouldn't you still want a condition here?

Reply:
No. In the previous step we verify that both have their contents as expected. If we skip docs, we won't publish docs below. Otherwise, we only want to publish after a successful build.

This breaks the test pipelines (as they won't be able to grab the artifact), but it's a much better experience than needing to entirely requeue the build if the `build` stage fails.

Reply:
@weshaggard I would love to override the artifacts on retry, but that's not really a thing with azure DevOps unfortunately.

Reply:
I think your continueOnFailure is really the best option.

Reply:
Went with resolving the artifact name at runtime. If it's a failed build, we set the artifact name to something with the build number+iteration.